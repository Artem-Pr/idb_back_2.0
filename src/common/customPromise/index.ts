export { CustomPromise } from './customPromise';
export { resolveAllSettled } from './resolveAllSettled';
