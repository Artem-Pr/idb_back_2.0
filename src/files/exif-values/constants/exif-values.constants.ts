export const EXIF_VALUES_CONSTANTS = {
  PAGINATION: {
    DEFAULT_PAGE: 1,
    DEFAULT_PER_PAGE: 50,
    MAX_PER_PAGE: 1000,
  },
  DATABASE: {
    COLLECTION_NAME: 'photos',
    EXIF_FIELD_PREFIX: 'exif.',
  },
  VALIDATION: {
    MIN_PROPERTY_NAME_LENGTH: 1,
    MAX_PROPERTY_NAME_LENGTH: 100,
  },
  PERFORMANCE: {
    QUERY_TIMEOUT_MS: 5000,
    MAX_VALUES_THRESHOLD: 100000,
  },
} as const;
