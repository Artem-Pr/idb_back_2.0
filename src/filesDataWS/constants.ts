export enum WebSocketActions {
  CREATE_PREVIEWS = 'CREATE_PREVIEWS',
  CREATE_PREVIEWS_STOP = 'CREATE_PREVIEWS_STOP',
  FILES_TEST = 'FILES_TEST',
  SYNC_PREVIEWS = 'SYNC_PREVIEWS',
  UNKNOWN_ACTION = 'UNKNOWN_ACTION',
  UPDATE_EXIF = 'UPDATE_EXIF',
  UPDATE_EXIF_STOP = 'UPDATE_EXIF_STOP',
}

export enum WSApiStatus {
  BUSY = 'busy',
  DONE = 'done',
  ERROR = 'error',
  INIT = 'init',
  PENDING = 'pending',
  PENDING_ERROR = 'pending-error',
  PENDING_SUCCESS = 'pending-success',
  READY = 'ready',
  STOPPED = 'stopped',
}
