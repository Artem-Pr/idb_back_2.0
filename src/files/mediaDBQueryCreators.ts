import { getEscapedString } from 'src/common/utils';
import type {
  GetFilesFiltersInputDto,
  GetFilesFoldersDto,
  GetFilesPaginationInputDto,
} from './dto/get-files-input.dto';
import type { Media } from './entities/media.entity';
import { ObjectId } from 'mongodb';
import type { SortingObject } from './types';
import { isEmpty } from 'ramda';

export type MongoFilterCondition = Partial<
  Record<keyof Media | '$or' | '$expr', object>
>;
type MongoAggregationFacet = Partial<
  Record<'response' | 'total' | 'pagination', object[]>
>;
export type MongoAggregationPipeline = Partial<
  | Record<
      '$match' | '$facet' | '$group' | '$project' | '$sort' | '$sample',
      object
    >
  | Record<'$unset', string | string[]>
>[];

interface MongoAggregationIncomingProps {
  conditions?: MongoFilterCondition[];
  facet?: MongoAggregationFacet;
  sorting?: SortingObject;
  sample?: { size: number };
}

type FilesFilter = Record<
  Exclude<keyof GetFilesFiltersInputDto, 'includeAllSearchTags'>,
  MongoFilterCondition
>;

export class MediaDBQueryCreators {
  ormFindByIdsQuery(ids: string[]) {
    return {
      where: {
        _id: { $in: ids.map((id) => new ObjectId(id)) },
      },
    };
  }

  getUpdateMediaQuery(updatedMedia: Media[]) {
    return updatedMedia.map((media) => ({
      updateOne: {
        filter: { _id: media._id },
        update: { $set: media },
      },
    }));
  }

  getReplacementMediaQuery(updatedMedia: Media[]) {
    return updatedMedia.map((media) => ({
      replaceOne: {
        filter: { _id: media._id },
        replacement: media,
      },
    }));
  }

  private getValidDateRange(dateRange: GetFilesFiltersInputDto['dateRange']) {
    if (dateRange) {
      const startDate = new Date(dateRange[0]);
      const endDate = new Date(dateRange[1]);

      return {
        $gte: startDate,
        $lt: endDate,
      };
    }
    return {};
  }

  getMongoFiltersConditions = (
    filesFilter: GetFilesFiltersInputDto,
  ): MongoFilterCondition[] => {
    const allFilters: FilesFilter = {
      fileName: {
        originalName: { $regex: filesFilter.fileName, $options: 'i' },
      },
      rating: { rating: { $eq: filesFilter.rating } },
      description: {
        description: { $regex: filesFilter.description, $options: 'i' },
      },
      anyDescription: { description: { $exists: true, $ne: '' } },
      searchTags: {
        keywords: {
          [filesFilter.includeAllSearchTags ? '$all' : '$in']:
            filesFilter.searchTags || [],
        },
      },
      excludeTags: { keywords: { $nin: filesFilter.excludeTags || [] } },
      mimetypes: {
        $or: filesFilter.mimetypes?.map((mimetype) => ({ mimetype })),
      },
      dateRange: {
        originalDate: this.getValidDateRange(filesFilter.dateRange),
      },
    };

    return Object.keys(allFilters)
      .filter((key) => filesFilter[key])
      .map((key) => allFilters[key]);
  };

  private getFolderPathExcludeSubFolderRegExp(
    folderPathEscaped: string,
  ): RegExp {
    return new RegExp(`^/${folderPathEscaped}/[^/]+\\.*$`);
  }

  private getMongoFilesExcludeFilesInSubfoldersCondition(
    folderPath: string,
  ): MongoFilterCondition {
    const folderPathEscaped = getEscapedString(folderPath);
    const folderPathExcludeSubFolderRegExp =
      this.getFolderPathExcludeSubFolderRegExp(folderPathEscaped);
    return { filePath: { $regex: folderPathExcludeSubFolderRegExp } };
  }

  getMongoFoldersCondition({
    folderPath,
    showSubfolders,
  }: GetFilesFoldersDto): MongoFilterCondition {
    if (folderPath && !showSubfolders) {
      return this.getMongoFilesExcludeFilesInSubfoldersCondition(folderPath);
    }
    if (folderPath && showSubfolders) {
      return {
        $expr: { $eq: [{ $indexOfCP: ['$filePath', `/${folderPath}/`] }, 0] },
      };
    }

    return {};
  }

  getMongoDynamicFoldersFacet(): MongoAggregationFacet {
    return {
      response: [
        { $project: { filePath: 1 } }, // includes only the filePath field in the resulting documents
        {
          $group: {
            _id: null,
            filePathSet: { $addToSet: '$filePath' }, // creates a set of unique file paths
          },
        },
        { $unset: ['_id', 'items'] }, // removes the _id field and the items field (if it exists) from the resulting documents
      ],
    };
  }

  getMongoFilesFacet({
    page,
    perPage,
  }: GetFilesPaginationInputDto): MongoAggregationFacet {
    return {
      response: [
        { $skip: page > 0 ? (page - 1) * perPage : 0 },
        { $limit: perPage },
      ],
      total: [
        { $group: { _id: null, filesSizeSum: { $sum: '$size' } } },
        { $unset: '_id' },
      ],
      pagination: [
        { $count: 'resultsCount' },
        {
          $addFields: {
            totalPages: {
              $ceil: {
                $divide: ['$resultsCount', perPage],
              },
            },
          },
        },
        {
          $addFields: {
            currentPage: {
              $cond: [{ $gt: [page, '$totalPages'] }, 1, page],
            },
          },
        },
      ],
    };
  }

  getMongoUsedKeywordsAggregation() {
    return [
      {
        $group: {
          _id: null,
          keywordsSet: { $addToSet: '$keywords' },
        },
      },
      {
        $project: {
          response: {
            $reduce: {
              input: { $concatArrays: '$keywordsSet' },
              initialValue: [],
              in: { $setUnion: ['$$value', '$$this'] },
            },
          },
        },
      },
      { $unset: '_id' },
    ];
  }

  private hasSorting(
    sorting?: MongoAggregationIncomingProps['sorting'],
    sample?: MongoAggregationIncomingProps['sample'],
  ) {
    return Boolean(!sample && sorting && !isEmpty(sorting));
  }

  createMongoAggregationPipeline({
    conditions,
    facet,
    sample,
    sorting,
  }: MongoAggregationIncomingProps): MongoAggregationPipeline {
    const aggregation: MongoAggregationPipeline = [];
    conditions?.length && aggregation.push({ $match: { $and: conditions } }); // conditionArr.length ? {$and: conditionArr} : {}
    this.hasSorting(sorting, sample) && aggregation.push({ $sort: sorting });
    sample && aggregation.push({ $sample: sample });
    facet && aggregation.push({ $facet: facet });

    return aggregation;
  }
}
