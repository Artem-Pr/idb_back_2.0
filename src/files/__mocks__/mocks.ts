import { Media } from '../entities/media.entity';
import { ObjectId } from 'mongodb';
import { MediaTemp } from '../entities/media-temp.entity';
import type { ProcessFile } from '../types';
import type { UploadFileOutputDto } from '../dto/upload-file-output.dto';
import { omit } from 'ramda';
import { SupportedImageMimetypes } from 'src/common/constants';
import { ExifDateTime } from 'exiftool-vendored';

const dateTimeOriginalMock = new ExifDateTime(
  2019,
  9,
  19,
  12,
  0,
  0,
  0,
  undefined,
  '2019:09:19 12:00:00',
);

export const exifDataMock = {
  DateTimeOriginal: dateTimeOriginalMock,
  Description: 'test description',
  GPSPosition: '42.5, 42.5',
  ImageSize: '1920x1080',
  Megapixels: 12,
  Rating: 5,
};

export const uploadFileMock: UploadFileOutputDto = {
  properties: {
    id: '665a0291b9e676b1015cf8b9',
    changeDate: null,
    duplicates: [
      {
        exif: exifDataMock,
        filePath: '/main/для теста базы/IMG_1728.heic',
        mimetype: SupportedImageMimetypes.heic,
        originalName: 'IMG_1728.heic',
        staticPreview:
          'http://localhost:3000/previews/image-heic/preview/2023.07.02 - originalDate/9eec89c3e7cf18920302f4e55d10aa52-preview.jpg',
        staticPath:
          'http://localhost:3000/previews/image-heic/fullSize/2023.07.02 - originalDate/9eec89c3e7cf18920302f4e55d10aa52-fullSize.jpg',
        staticVideoFullSize: null,
      },
      {
        exif: exifDataMock,
        filePath: '/main/SD/IMG_1728.heic',
        mimetype: SupportedImageMimetypes.heic,
        originalName: 'IMG_1728.heic',
        staticPreview:
          'http://localhost:3000/previews/image-heic/preview/2023.07.02 - originalDate/f7d6132c59acb526c1df74f438c744fa-preview.jpg',
        staticPath:
          'http://localhost:3000/previews/image-heic/fullSize/2023.07.02 - originalDate/f7d6132c59acb526c1df74f438c744fa-fullSize.jpg',
        staticVideoFullSize: null,
      },
    ],
    description: exifDataMock.Description,
    exif: exifDataMock,
    filePath: null,
    imageSize: '2268x4032',
    keywords: [],
    megapixels: 9.1,
    mimetype: SupportedImageMimetypes.heic,
    originalDate: new Date('2023-07-02T17:36:33.000Z'),
    originalName: 'IMG_1728.heic',
    rating: 4,
    size: 1950900,
    staticPath:
      'http://localhost:3000/previews/5136bc14-512f-47f9-a551-304bb254f528-fullSize.jpg',
    staticPreview:
      'http://localhost:3000/temp/5136bc14-512f-47f9-a551-304bb254f528-preview.jpg',
    staticVideoFullSize: null,
    timeStamp: '00:00:00.000',
  },
};

export class UploadFileMock {
  _properties: UploadFileOutputDto['properties'] = uploadFileMock.properties;

  constructor({ properties }: Partial<UploadFileOutputDto> = {}) {
    properties && (this.properties = properties);
  }

  set exif(exif: UploadFileOutputDto['properties']['exif']) {
    this._properties.exif = exif;
  }
  addExif(exif: UploadFileOutputDto['properties']['exif']) {
    this._properties.exif = { ...this._properties.exif, ...exif };
  }

  set properties(properties: UploadFileOutputDto['properties']) {
    this._properties = properties;
  }
  addProperties(properties: Partial<UploadFileOutputDto['properties']>) {
    this._properties = { ...this._properties, ...properties };
  }

  set duplicates(duplicates: UploadFileOutputDto['properties']['duplicates']) {
    this._properties.duplicates = duplicates;
  }

  addDuplicates(duplicates: UploadFileOutputDto['properties']['duplicates']) {
    this._properties.duplicates = [
      ...this._properties.duplicates,
      ...duplicates,
    ];
  }

  set staticPath(staticPath: UploadFileOutputDto['properties']['staticPath']) {
    this._properties.staticPath = staticPath;
  }

  set staticPreview(
    staticPreview: UploadFileOutputDto['properties']['staticPreview'],
  ) {
    this._properties.staticPreview = staticPreview;
  }

  updateFromMedia(media: Media) {
    if (media.exif) {
      this.addExif(media.exif);
    }

    const properties: UploadFileOutputDto['properties'] = {
      id: media._id.toString(),
      duplicates: [],
      filePath: null,
      staticPath: this._properties.staticPath,
      staticPreview: this._properties.staticPreview,
      staticVideoFullSize: null,
      ...omit(['filePath', 'preview', 'fullSizeJpg', '_id'], media),
    };

    this.properties = {
      ...this._properties,
      ...properties,
    };
  }

  get uploadFile(): UploadFileOutputDto {
    return {
      properties: this._properties,
    };
  }
}

export function createMockProcessFile(): ProcessFile {
  const mock: ProcessFile = {
    filename: 'mockFile.jpg',
    mimetype: SupportedImageMimetypes.jpg,
    originalname: 'original_mock_file.jpg',
    size: 1024,
  };

  return mock;
}

export function createMediaMock({
  id = new ObjectId(),
  name = 'mockFile',
  originalNameWithoutExt = 'original_mock_file',
}: {
  id?: ObjectId;
  name?: string;
  originalNameWithoutExt?: string;
} = {}): Media {
  const media = new Media();

  media._id = id;
  media.originalName = `${originalNameWithoutExt}.jpg`;
  media.mimetype = SupportedImageMimetypes.jpg;
  media.size = 1024;
  media.megapixels = exifDataMock.Megapixels;
  media.imageSize = exifDataMock.ImageSize;
  media.keywords = ['test', 'media'];
  media.changeDate = new Date('2023-07-02T17:36:33.000Z').getTime();
  media.originalDate = exifDataMock.DateTimeOriginal.toDate();
  media.filePath = `/path/to/${name}.jpg`;
  media.preview = `/path/to/${name}-preview.jpg`;
  media.fullSizeJpg = `/path/to/${name}-fullSize.jpg`;
  media.rating = exifDataMock.Rating;
  media.description = exifDataMock.Description;
  media.timeStamp = '00:00:00.100';
  media.exif = exifDataMock;

  return media;
}

export function createMediaTempMock({
  id = new ObjectId(),
  name = 'mockTempFile',
  originalNameWithoutExt = 'original_mock_temp_file',
}: {
  id?: ObjectId;
  name?: string;
  originalNameWithoutExt?: string;
} = {}): MediaTemp {
  const mediaTemp = createMediaMock({
    id,
    name,
    originalNameWithoutExt,
  });
  return mediaTemp;
}
