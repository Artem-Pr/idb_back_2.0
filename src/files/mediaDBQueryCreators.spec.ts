import type { MongoFilterCondition } from './mediaDBQueryCreators';
import { MediaDBQueryCreators } from './mediaDBQueryCreators';
import { ObjectId } from 'mongodb';
import type {
  GetFilesFiltersInputDto,
  GetFilesFoldersDto,
  GetFilesPaginationInputDto,
} from './dto/get-files-input.dto';
import type { SortingObject } from './types';
import { SupportedImageMimetypes } from 'src/common/constants';
import { createMediaMock } from './__mocks__/mocks';
import type { Media } from './entities/media.entity';

describe('MediaDBQueryCreators', () => {
  let mediaDBQueryCreators: MediaDBQueryCreators;

  beforeEach(() => {
    mediaDBQueryCreators = new MediaDBQueryCreators();
  });

  describe('ormFindByIdsQuery', () => {
    it('should return a query object with ObjectId array', () => {
      const ids = ['507f1f77bcf86cd799439011', '507f1f77bcf86cd799439012'];
      const query = mediaDBQueryCreators.ormFindByIdsQuery(ids);
      expect(query).toEqual({
        where: {
          _id: {
            $in: [new ObjectId(ids[0]), new ObjectId(ids[1])],
          },
        },
      });
    });
  });

  describe('getUpdateMediaQuery', () => {
    it('should return update media query object', () => {
      const mediaMock1: Media = createMediaMock({ name: 'mediaMock1' });
      const mediaMock2: Media = createMediaMock({ name: 'mediaMock2' });

      const query = mediaDBQueryCreators.getUpdateMediaQuery([
        mediaMock1,
        mediaMock2,
      ]);
      expect(query).toEqual([
        {
          updateOne: {
            filter: { _id: mediaMock1._id },
            update: { $set: mediaMock1 },
          },
        },
        {
          updateOne: {
            filter: { _id: mediaMock2._id },
            update: { $set: mediaMock2 },
          },
        },
      ]);
    });
  });

  describe('getValidDateRange', () => {
    it('should return date range object', () => {
      const dateRange: [string, string] = ['2022-01-01', '2022-12-31'];
      const result = mediaDBQueryCreators['getValidDateRange'](dateRange);
      expect(result).toEqual({
        $gte: new Date(dateRange[0]),
        $lt: new Date(dateRange[1]),
      });
    });

    it('should return an empty object when dateRange is undefined', () => {
      const result = mediaDBQueryCreators['getValidDateRange'](undefined);
      expect(result).toEqual({});
    });
  });

  describe('getMongoFiltersConditions', () => {
    it('should return an array of Mongo filter conditions', () => {
      const filesFilter: GetFilesFiltersInputDto = {
        fileName: 'test',
        rating: 5,
        description: 'test description',
        includeAllSearchTags: true,
        searchTags: ['tag1'],
        excludeTags: ['tag2'],
        mimetypes: [SupportedImageMimetypes.jpeg],
        dateRange: ['2022-01-01', '2022-12-31'],
      };

      const result =
        mediaDBQueryCreators.getMongoFiltersConditions(filesFilter);
      expect(result).toEqual([
        { originalName: { $regex: 'test', $options: 'i' } },
        { rating: { $eq: 5 } },
        { description: { $regex: 'test description', $options: 'i' } },
        { keywords: { $all: ['tag1'] } },
        { keywords: { $nin: ['tag2'] } },
        { $or: [{ mimetype: 'image/jpeg' }] },
        {
          originalDate: {
            $gte: new Date('2022-01-01'),
            $lt: new Date('2022-12-31'),
          },
        },
      ]);
    });
  });

  describe('getFolderPathExcludeSubFolderRegExp', () => {
    it('should return a regular expression to exclude subfolders', () => {
      const regExp =
        mediaDBQueryCreators['getFolderPathExcludeSubFolderRegExp']('path');
      expect(regExp).toEqual(/^\/path\/[^/]+\.*$/);
    });
  });

  describe('getMongoFilesExcludeFilesInSubfoldersCondition', () => {
    it('should return a Mongo filter condition to exclude files in subfolders', () => {
      const condition =
        mediaDBQueryCreators['getMongoFilesExcludeFilesInSubfoldersCondition'](
          'path',
        );
      expect(condition).toEqual({ filePath: { $regex: /^\/path\/[^/]+\.*$/ } });
    });
  });

  describe('getMongoFoldersCondition', () => {
    it('should return a Mongo filter condition for folders without subfolders', () => {
      const folderFilter: GetFilesFoldersDto = {
        folderPath: 'path',
        showSubfolders: false,
      };
      const condition =
        mediaDBQueryCreators.getMongoFoldersCondition(folderFilter);
      expect(condition).toEqual({ filePath: { $regex: /^\/path\/[^/]+\.*$/ } });
    });

    it('should return a Mongo filter condition for folders with subfolders', () => {
      const folderFilter: GetFilesFoldersDto = {
        folderPath: 'path',
        showSubfolders: true,
      };
      const condition =
        mediaDBQueryCreators.getMongoFoldersCondition(folderFilter);
      expect(condition).toEqual({
        $expr: { $eq: [{ $indexOfCP: ['$filePath', '/path/'] }, 0] },
      });
    });

    it('should return an empty object when folderFilter is undefined', () => {
      const condition = mediaDBQueryCreators.getMongoFoldersCondition({});
      expect(condition).toEqual({});
    });
  });

  describe('getMongoDynamicFoldersFacet', () => {
    it('should return the dynamic folders facet', () => {
      const facet = mediaDBQueryCreators.getMongoDynamicFoldersFacet();
      expect(facet).toEqual({
        response: [
          { $project: { filePath: 1 } },
          { $group: { _id: null, filePathSet: { $addToSet: '$filePath' } } },
          { $unset: ['_id', 'items'] },
        ],
      });
    });
  });

  describe('getMongoFilesFacet', () => {
    it('should return the files facet', () => {
      const pagination: GetFilesPaginationInputDto = { page: 1, perPage: 10 };
      const facet = mediaDBQueryCreators.getMongoFilesFacet(pagination);
      expect(facet).toEqual({
        response: [{ $skip: 0 }, { $limit: 10 }],
        total: [
          { $group: { _id: null, filesSizeSum: { $sum: '$size' } } },
          { $unset: '_id' },
        ],
        pagination: [
          { $count: 'resultsCount' },
          {
            $addFields: {
              totalPages: { $ceil: { $divide: ['$resultsCount', 10] } },
            },
          },
          {
            $addFields: {
              currentPage: { $cond: [{ $gt: [1, '$totalPages'] }, 1, 1] },
            },
          },
        ],
      });
    });
  });

  describe('getMongoUsedKeywordsAggregation', () => {
    it('should return the used keywords aggregation', () => {
      const aggregation =
        mediaDBQueryCreators.getMongoUsedKeywordsAggregation();
      expect(aggregation).toEqual([
        { $group: { _id: null, keywordsSet: { $addToSet: '$keywords' } } },
        {
          $project: {
            response: {
              $reduce: {
                input: { $concatArrays: '$keywordsSet' },
                initialValue: [],
                in: { $setUnion: ['$$value', '$$this'] },
              },
            },
          },
        },
        { $unset: '_id' },
      ]);
    });
  });

  describe('createMongoAggregationPipeline', () => {
    const conditions: MongoFilterCondition[] = [
      { filePath: { $regex: /.*\.jpg$/ } },
    ];
    const facet = { response: [{ $skip: 0 }] };
    const sorting: SortingObject = { rating: 1 };
    const sample = { size: 10 };

    it('should create a Mongo aggregation pipeline with sorting', () => {
      const pipeline = mediaDBQueryCreators.createMongoAggregationPipeline({
        conditions,
        facet,
        sorting,
      });

      expect(pipeline).toEqual([
        { $match: { $and: conditions } },
        { $sort: sorting },
        { $facet: facet },
      ]);
    });

    it('should create a Mongo aggregation pipeline with sample', () => {
      const pipeline = mediaDBQueryCreators.createMongoAggregationPipeline({
        conditions,
        facet,
        sorting,
        sample,
      });

      expect(pipeline).toEqual([
        { $match: { $and: conditions } },
        { $sample: sample },
        { $facet: facet },
      ]);
    });

    it('should create a Mongo aggregation pipeline without sorting and sample', () => {
      const conditions: MongoFilterCondition[] = [
        { filePath: { $regex: /.*\.jpg$/ } },
      ];
      const facet = { response: [{ $skip: 0 }] };

      const pipeline = mediaDBQueryCreators.createMongoAggregationPipeline({
        conditions,
        facet,
      });

      expect(pipeline).toEqual([
        { $match: { $and: conditions } },
        { $facet: facet },
      ]);
    });
  });
});
