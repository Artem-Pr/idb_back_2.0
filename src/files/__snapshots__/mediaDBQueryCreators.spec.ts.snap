// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaDBQueryCreators getMongoEmptyFullSizesCondition should return correct condition with folderPath 1`] = `
{
  "$and": [
    {
      "$expr": {
        "$eq": [
          {
            "$indexOfCP": [
              "$filePath",
              "/path/to/folder/",
            ],
          },
          0,
        ],
      },
    },
    {
      "mimetype": {
        "$in": [
          "video/x-msvideo",
          "video/avi",
          "video/x-flv",
          "video/x-matroska",
          "video/quicktime",
          "video/mp4",
          "video/mpeg",
          "video/ogg",
          "video/webm",
          "video/x-ms-wmv",
          "image/heic",
        ],
      },
    },
    {
      "$or": [
        {
          "fullSizeJpg": {
            "$exists": false,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": null,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoEmptyFullSizesCondition should return correct condition with mimeTypes 1`] = `
{
  "$and": [
    {},
    {
      "mimetype": {
        "$in": [
          "video/mp4",
        ],
      },
    },
    {
      "$or": [
        {
          "fullSizeJpg": {
            "$exists": false,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": null,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoEmptyFullSizesCondition should return correct condition with mimeTypes and folderPath 1`] = `
{
  "$and": [
    {
      "$expr": {
        "$eq": [
          {
            "$indexOfCP": [
              "$filePath",
              "/path/to/folder/",
            ],
          },
          0,
        ],
      },
    },
    {
      "mimetype": {
        "$in": [
          "video/mp4",
        ],
      },
    },
    {
      "$or": [
        {
          "fullSizeJpg": {
            "$exists": false,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": null,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoEmptyFullSizesCondition should return correct condition without mimeTypes and folderPath 1`] = `
{
  "$and": [
    {},
    {
      "mimetype": {
        "$in": [
          "video/x-msvideo",
          "video/avi",
          "video/x-flv",
          "video/x-matroska",
          "video/quicktime",
          "video/mp4",
          "video/mpeg",
          "video/ogg",
          "video/webm",
          "video/x-ms-wmv",
          "image/heic",
        ],
      },
    },
    {
      "$or": [
        {
          "fullSizeJpg": {
            "$exists": false,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": null,
          },
        },
        {
          "fullSizeJpg": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoEmptyPreviewsCondition should return correct condition with folderPath 1`] = `
{
  "$and": [
    {
      "$expr": {
        "$eq": [
          {
            "$indexOfCP": [
              "$filePath",
              "/path/to/folder/",
            ],
          },
          0,
        ],
      },
    },
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "image/jpg",
          "image/png",
          "image/heic",
          "image/gif",
          "image/x-adobe-dng",
          "video/x-msvideo",
          "video/avi",
          "video/x-flv",
          "video/x-matroska",
          "video/quicktime",
          "video/mp4",
          "video/mpeg",
          "video/ogg",
          "video/webm",
          "video/x-ms-wmv",
        ],
      },
    },
    {
      "$or": [
        {
          "preview": {
            "$exists": false,
          },
        },
        {
          "preview": {
            "$eq": null,
          },
        },
        {
          "preview": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoEmptyPreviewsCondition should return correct condition with mimeTypes 1`] = `
{
  "$and": [
    {},
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "video/mp4",
        ],
      },
    },
    {
      "$or": [
        {
          "preview": {
            "$exists": false,
          },
        },
        {
          "preview": {
            "$eq": null,
          },
        },
        {
          "preview": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoEmptyPreviewsCondition should return correct condition with mimeTypes and folderPath 1`] = `
{
  "$and": [
    {
      "$expr": {
        "$eq": [
          {
            "$indexOfCP": [
              "$filePath",
              "/path/to/folder/",
            ],
          },
          0,
        ],
      },
    },
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "video/mp4",
        ],
      },
    },
    {
      "$or": [
        {
          "preview": {
            "$exists": false,
          },
        },
        {
          "preview": {
            "$eq": null,
          },
        },
        {
          "preview": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoEmptyPreviewsCondition should return correct condition without mimeTypes and folderPath 1`] = `
{
  "$and": [
    {},
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "image/jpg",
          "image/png",
          "image/heic",
          "image/gif",
          "image/x-adobe-dng",
          "video/x-msvideo",
          "video/avi",
          "video/x-flv",
          "video/x-matroska",
          "video/quicktime",
          "video/mp4",
          "video/mpeg",
          "video/ogg",
          "video/webm",
          "video/x-ms-wmv",
        ],
      },
    },
    {
      "$or": [
        {
          "preview": {
            "$exists": false,
          },
        },
        {
          "preview": {
            "$eq": null,
          },
        },
        {
          "preview": {
            "$eq": "",
          },
        },
      ],
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoFilesCondition should return correct condition with folderPath 1`] = `
{
  "$and": [
    {
      "$expr": {
        "$eq": [
          {
            "$indexOfCP": [
              "$filePath",
              "/path/to/folder/",
            ],
          },
          0,
        ],
      },
    },
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "image/jpg",
          "image/png",
          "image/heic",
          "image/gif",
          "image/x-adobe-dng",
          "video/x-msvideo",
          "video/avi",
          "video/x-flv",
          "video/x-matroska",
          "video/quicktime",
          "video/mp4",
          "video/mpeg",
          "video/ogg",
          "video/webm",
          "video/x-ms-wmv",
        ],
      },
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoFilesCondition should return correct condition with mimeTypes 1`] = `
{
  "$and": [
    {},
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "video/mp4",
        ],
      },
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoFilesCondition should return correct condition with mimeTypes and folderPath 1`] = `
{
  "$and": [
    {
      "$expr": {
        "$eq": [
          {
            "$indexOfCP": [
              "$filePath",
              "/path/to/folder/",
            ],
          },
          0,
        ],
      },
    },
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "video/mp4",
        ],
      },
    },
  ],
}
`;

exports[`MediaDBQueryCreators getMongoFilesCondition should return correct condition without mimeTypes and folderPath 1`] = `
{
  "$and": [
    {},
    {
      "mimetype": {
        "$in": [
          "image/jpeg",
          "image/jpg",
          "image/png",
          "image/heic",
          "image/gif",
          "image/x-adobe-dng",
          "video/x-msvideo",
          "video/avi",
          "video/x-flv",
          "video/x-matroska",
          "video/quicktime",
          "video/mp4",
          "video/mpeg",
          "video/ogg",
          "video/webm",
          "video/x-ms-wmv",
        ],
      },
    },
  ],
}
`;
